### PR TITLE
build.d: Reduce the default number of jobs in cgroups

### PR DESCRIPTION
**Pull request by tim-dlang**
CircleCI runs the tests on a server with many CPUs, but restricts the
number of CPUs and the memory using Linux cgroups. Using the total
number of CPUs as the default number of jobs can result in many
parallel DMD processes, which can consume too much memory. This can
result in random failures.

This commit tries to detect a reduced number of CPUs, so the number of
jobs can be decreased.

### CODE REVIEW

**Comment by tim-dlang** [comment]
From the [CircleCI documentation](https://circleci.com/docs/configuration-reference/#resourceclass):
> Java, Erlang and any other languages that introspect the /proc directory for information about CPU count may require additional configuration to prevent them from slowing down when using the CircleCI resource class feature. Programs with this issue may request 32 CPU cores and run slower than they would when requesting one core. Users of languages with this issue should pin their CPU count to their guaranteed CPU resources.

The program build.d uses totalCPUs to determine the number of jobs. The log on CircleCI for this pull request shows 36. This could explain random failures like for https://github.com/dlang/dlang.org/pull/3681, because the jobs can run out of memory.

File: compiler/src/build.d [hunk 1]
```diff
@@ -78,6 +78,30 @@ int main(string[] args)
...
+        {
+            string cpuSharesStr = readText(cpuSharesFilename).strip;
+            try
+            {
+                int cpuShares = cpuSharesStr.to!int;
+                jobs = min(jobs, max(1, cpuShares / 1024));
+            }
+            catch (ConvException)
+            {
+                stderr.writeln("Warning: /sys/fs/cgroup/cpu/cpu.shares contains unknown value:", cpuSharesStr);
```
Comment (dkorpel) [inline compiler/src/build.d:101]: This is an uninformative warning message. It should tell what it was looking for, what the consequence of the failure is, and suggest how to fix it.

**Comment by Imperatorn** [comment]
Any idea why it reports 36 instead of 32?

**Comment by tim-dlang** [comment]
> Any idea why it reports 36 instead of 32?

I don't know exactly. Maybe they had only servers with 32 CPU cores when the documentation was written. Later they could have added servers with 36 cores. If they have different servers, that could also further explain, why the tests only sometimes run out of memory.

**Comment by tim-dlang** [comment]
Closing in favour of https://github.com/dlang/dmd/pull/15799 and https://github.com/dlang/dlang.org/pull/3724.

**Pull request closed without merging**